lua/rainbow-delimiters/strategy/global.lua

```diff
@@ -64,36 +78,63 @@ local function update_range(bufnr, changes, tree, lang)
 	local matches = Stack.new()
 
 	for _, change in ipairs(changes) do
+		-- This is the match record, it lists all the relevant nodes from
+		-- each match.
+		local match_record = new_match_record()
 		local root_node = tree:root()
 		local start_row, end_row = change[1], change[3] + 1
 		lib.clear_namespace(bufnr, lang, start_row, end_row)
-		for _, match, _ in query:iter_matches(root_node, bufnr, start_row, end_row) do
-			-- This is the match record, it lists all the relevant nodes from
-			-- the match.
-			local match_record = {
-				opening = Stack.new(),
-				closing = Stack.new(),
-				intermediate = Stack.new(),
-				children = Stack.new(),
-			}
-			for id, node in pairs(match) do
-				local name = query.captures[id]
-				if name == 'container' then
-					match_record.container = node
-				else
-					if match_record[name] then match_record[name]:push(node) end
+		for qid, node, _ in query:iter_captures(root_node, bufnr, start_row, end_row) do
+			local name = query.captures[qid]
+			-- check for 'delimiter' first, since that should be the most
+			-- common name
+			if name == 'delimiter' then
+				match_record.delimiter:push(node)
+			elseif name == 'container' and not match_record.container then
+				match_record.container = node
+			elseif name == 'container' then
+				local prev_closed = true
+				if not match_record.sentinel then
+					-- remember if we have closed this match_record
+					prev_closed = false
 				end
-			end
-
-			for _, other in matches:iter() do
-				if not ts.is_ancestor(match_record.container, other.container) then
-					break
+				matches:push(match_record)
+				match_record = new_match_record(node)
+				if not prev_closed then
+					-- if we didn't close the previous match_record, it must
+					-- mean that the current match_record has it as an ancestor
+					match_record.has_ancestor = true
 				end
-				match_record.children:push(other)
-				matches:pop()
+			elseif name == 'sentinel' then
+				match_record.sentinel = true
+				-- if we see the sentinel, then we are done with the current
+				-- container
+				if match_record.has_ancestor then
+					local prev_match_record = matches:pop()
+					if prev_match_record then
+						-- since we have an ancestor, it has to be the last
+						-- element of the stack
+						prev_match_record.children:push(match_record)
+						match_record = prev_match_record
+					else
+						-- since match_record.has_ancestor was true, we shouldn't
+						-- be able to get to here unless something went wrong
+						-- with the queries or treesitter itself
+						log.debug([[You are missing a @container,
+									which should be impossible!
+									Please double check the queries.]])
+					end
+				else
+					-- if match_record doesn't have an ancestor, the sentinel
+					-- means that we are done with it
+					matches:push(match_record)
+					match_record = new_match_record()
```

Comment:
This can be changed to
```suggestion
					match_record = nil
```
if we change the above code as commented. This might be make the code easier to follow?